pyinterprod/interpro/contrib/ncbifam.py

```
@@ -54,3 +63,90 @@ def get_signatures(hmm_file: str, info_file: str):
+    list_ids = []
+    retmax = 10000
+    if int(total_ids) > retmax:
+        for retstart in range(0, int(total_ids), retmax):
+            r = _fetch_url_xml(f"{BASE_IDS_URL}&retmax={retmax}&retstart={retstart}")
+            for i in r.findall('./IdList/Id'):
+                list_ids.append(i.text)
+    else:
+        r = _fetch_url_xml(f"{BASE_IDS_URL}&retmax={total_ids}")
+        for i in r.findall('./IdList/Id'):
+            list_ids.append(i.text)
+    return set(list_ids)
+
+
+def get_all_accessions_list(ids_list: list) -> set:
+    ncbi_accessions_list = []
+    step = 200
+    for i in range(0, len(ids_list), step):
+        r = _fetch_url_xml(f"{BASE_ACCESSIONS_URL}&id={','.join(ids_list[i:i+step])}")
+        accession = r.findall('./DocumentSummarySet/DocumentSummary/DispFamilyAcc')
+        for a in accession:
+            ncbi_accessions_list.append(a.text.split(".")[0])
+    return set(ncbi_accessions_list)
+
+
+def get_ncbifam_info(accessions_list: set):
+    url_list = []
+    for accession in accessions_list:
+        url_list.append(f"{BASE_NCBIFAM_URL}&match=accession_._{accession}")
+
+    result_list = []
+    total_accessions_count = len(accessions_list)
+    total_complete_requests = 0
+    with ThreadPoolExecutor(max_workers=8) as executor:
+        futures_to_data = {executor.submit(request.urlopen, i): i for i in url_list}
+        while total_complete_requests < total_accessions_count:
+            for future in as_completed(futures_to_data):
+                if future.exception():
+                    data = _retry(future, futures_to_data, executor)
+                    logger.info(f'Failure, retrying {data}')
+                else:
+                    result = future.result().read()
+                    filtered_info = _filter_ncbifam_info(result.decode('utf-8'))
+                    result_list.append(filtered_info)
+                    total_complete_requests += 1
+                futures_to_data.pop(future)
```

A few things can be improved/simplified here:

- I don't think it's necessary to keep track of the number of successful tasks. `while futures_to_data` should be enough.
- We should expect a non-significant number of failures. To avoid drowning in log messages, you remove the call to `logger.info()`.
- However, we only want ignore errors and retry, if the request fails because we already submitted too many requests (in which case, the response would have a 429 HTTP code). If the request fails for any other reason, we should just stop the script.
- I know it's good practice to define small functions, but `_retry()` is so small I don't think it's necessary to keep its content in a separate function.